AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -3,43 +3,129 @@
-          if (tokenResponsejsonError) {
-            MSLogError([MSDataStore logTag], @"Can't deserialize tokens with error: %@", [tokenResponsejsonError description]);
-            completion([[MSTokensResponse alloc] initWithTokens:nil], error);
-          }
-
-          // Create token result object.
-          MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
-          completion(tokens, error);
++ (MSTokenResult *)retrieveCachedToken:(NSString *)partitionName {
+    NSString *tokenString = [MSKeychainUtil stringForKey:[MSTokenExchange tokenKeyNameForPartition:partitionName]];
+    if(tokenString != nil){
+        MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithString:tokenString];
+        
+        if(tokenResult != nil){
+            
+            NSDate *currentUTCDate = [NSDate date];
+            
+            NSDateFormatter* utcFormatter = [[NSDateFormatter alloc] init];
+            [utcFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
+            NSDate *tokenExpireDate = [utcFormatter dateFromString:tokenResult.expiresOn];
+            
+            if([currentUTCDate laterDate:tokenExpireDate] == currentUTCDate){
+                [MSTokenExchange removeCachedToken:partitionName];
+                
+                 MSLogWarning([MSDataStore logTag], @"The token in the cache has expired for the partitionKey : %@.", partitionName);
+                return nil;
+            }
+            
+            MSLogDebug([MSDataStore logTag], @"Retrieved token from keychain for the partitionKey : %@.", partitionName);
+            return tokenResult;
         }
-      }];
+    }
+    
+     MSLogWarning([MSDataStore logTag], @"Failed to retrieve token from keychain or none was found for the partitionKey : %@.", partitionName);
+    
+    return nil;
+}
+
++ (void)removeCachedToken:(NSString *)partitionName {
+    NSString *tokenString = [MSKeychainUtil deleteStringForKey:[MSTokenExchange tokenKeyNameForPartition:partitionName]];
+    
+    if (tokenString) {
+        MSLogDebug([MSDataStore logTag], @"Removed token from keychain for the partitionKey : %@.", partitionName);
+    } else {
+        MSLogWarning([MSDataStore logTag], @"Failed to remove token from keychain or none was found for the partitionKey : %@.", partitionName);
+    }
+}
+
++ (void)removeAllCachedTokens {
+    NSString *readonlyTokenString = [MSKeychainUtil deleteStringForKey:kMSStorageReadOnlyDbTokenKey];
+    NSString *userTokenString = [MSKeychainUtil deleteStringForKey:kMSStorageUserDbTokenKey];
+    if (readonlyTokenString && userTokenString) {
+        MSLogDebug([MSDataStore logTag], @"Removed all the tokens from keychain.");
+    } else {
+        MSLogWarning([MSDataStore logTag], @"Failed to remove all of the tokens from keychain");
+    }
+}
+
++ (NSString *)tokenKeyNameForPartition:(NSString *)partitionName{
+    
+    NSString *tokenKeyName = kMSStorageReadOnlyDbTokenKey;
```

Comment:
The method can be simplified as below.
```
    if ([partitionName containsString:MSDataStoreAppDocumentsPartition]) {
        return kMSStorageReadOnlyDbTokenKey;
    }

    return kMSStorageUserDbTokenKey;
```
You don't need `!` in condition if you reverse the result, you can get rid of one unnecessary bit operation.